File: src-tauri/src/dns.rs
use crate::runner::{run_elevated_script, write_ps1};
use serde::Serialize;
use std::net::ToSocketAddrs;
use std::path::Path;

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DnsProvider {
    pub id: &'static str,
    pub name: &'static str,
    pub primary: &'static str,
    pub secondary: &'static str,
    pub doh_template: &'static str,
    /// Короткая подпись рядом с названием.
    pub note: &'static str,
    /// Подробное описание (показывается в раскрывающемся блоке).
    pub description: &'static str,
}

const PROVIDERS: &[DnsProvider] = &[
    DnsProvider {
        id: "google",
        name: "Google Public DNS",
        primary: "8.8.8.8",
        secondary: "8.8.4.4",
        doh_template: "https://dns.google/dns-query",
        note: "крупный Anycast",
        description: "Публичный DNS Google. Очень стабильный и быстрый, без фильтрации. \
                      Минус — принадлежит Google (логи). Хороший выбор по умолчанию.",
    },
    DnsProvider {
        id: "cloudflare",
        name: "Cloudflare",
        primary: "1.1.1.1",
        secondary: "1.0.0.1",
        doh_template: "https://cloudflare-dns.com/dns-query",
        note: "быстрый, минимум логов",
        description: "Заявлен как самый быстрый публичный DNS, логов почти не ведёт. \
                      Без фильтрации. Хорош для скорости и приватности.",
    },
    DnsProvider {
        id: "adguard",
        name: "AdGuard DNS",
        primary: "94.140.14.14",
        secondary: "94.140.15.15",
        doh_template: "https://dns.adguard-dns.com/dns-query",
        note: "блокирует рекламу и трекеры",
        description: "Блокирует рекламу, трекеры и часть вредоносных доменов на уровне DNS. \
                      Может резать и нужные сайты — если что-то не открывается, проверьте здесь.",
    },
    DnsProvider {
        id: "xbox-dns",
        name: "XBox DNS",
        primary: "111.88.96.50",
        secondary: "111.88.96.51",
        doh_template: "https://xbox-dns.ru/dns-query",
        note: "Smart DNS: Xbox Live, ИИ, игры",
        description: "Smart DNS для доступа к сервисам с региональными ограничениями: \
                      Xbox Live (ошибка 0x80a40401), игры Supercell (Brawl Stars, Clash of Clans), \
                      ChatGPT/Gemini/Claude/Copilot, JetBrains/GitHub/Notion, Twitch/Spotify/xCloud. \
                      Обычный трафик идёт напрямую, только гео-ограниченные — через шлюз. \
                      Без регистрации, заявлены DoH/DoT и Zero-Log.",
    },
    DnsProvider {
        id: "quad9",
        name: "Quad9",
        primary: "9.9.9.9",
        secondary: "149.112.112.112",
        doh_template: "https://dns.quad9.net/dns-query",
        note: "защита от вредоносных доменов",
        description: "Блокирует вредоносные и фишинговые домены (по базе угроз). \
                      Без рекламных блокировок. Некоммерческий, ориентирован на безопасность.",
    },
    DnsProvider {
        id: "opendns",
        name: "OpenDNS (Cisco)",
        primary: "208.67.222.222",
        secondary: "208.67.220.220",
        doh_template: "https://doh.opendns.com/dns-query",
        note: "стабильный, есть фильтры Cisco",
        description: "DNS от Cisco. Стабильный, с опциональной фильтрацией (по умолчанию — \
                      базовая защита от фишинга). Хорош как надёжный резерв.",
    },
    DnsProvider {
        id: "dns-sb",
        name: "DNS.SB",
        primary: "185.222.222.222",
        secondary: "45.11.45.11",
        doh_template: "https://doh.dns.sb/dns-query",
        note: "независимый, без логов",
        description: "Независимый публичный DNS, без логов и фильтрации. Хорошая альтернатива \
                      большим корпоративным резолверам.",
    },
    DnsProvider {
        id: "mullvad",
        name: "Mullvad DNS",
        primary: "194.242.2.2",
        secondary: "194.242.2.3",
        doh_template: "https://doh.mullvad.net/dns-query",
        note: "приватный, без логов",
        description: "DNS от Mullvad, известных своим отношением к приватности: \
                      без логов, без фильтрации. Доступен без VPN-подписки.",
    },
];

pub fn providers() -> &'static [DnsProvider] {
    PROVIDERS
}

pub fn provider(id: &str) -> Option<&'static DnsProvider> {
    PROVIDERS.iter().find(|p| p.id == id)
}

// ------------------------------------------------------------- benchmark

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DnsPing {
    pub id: String,
    pub primary_ms: Option<u64>,
    pub secondary_ms: Option<u64>,
    /// Среднее по обоим адресам (для сортировки/сравнения).
    pub avg_ms: Option<u64>,
    pub error: Option<String>,
}

/// Строит минимальный DNS-запрос (A-запись) для домена.
fn build_dns_query(domain: &str) -> Vec<u8> {
    let mut q = Vec::new();
    q.extend_from_slice(&[0x12, 0x34]); // ID
    q.extend_from_slice(&[0x01, 0x00]); // flags: standard query, RD
    q.extend_from_slice(&[0x00, 0x01]); // QDCOUNT
    q.extend_from_slice(&[0x00, 0x00]); // ANCOUNT
    q.extend_from_slice(&[0x00, 0x00]); // NSCOUNT
    q.extend_from_slice(&[0x00, 0x00]); // ARCOUNT
    for label in domain.split('.') {
        q.push(label.len() as u8);
        q.extend_from_slice(label.as_bytes());
    }
    q.push(0);
    q.extend_from_slice(&[0x00, 0x01]); // QTYPE A
    q.extend_from_slice(&[0x00, 0x01]); // QCLASS IN
    q
}

/// Один UDP DNS-запрос: возвращает время отклика в мс (или None).
fn query_once(server: &str, domain: &str, timeout: std::time::Duration) -> Option<u64> {
    use std::net::UdpSocket;
    let sock = UdpSocket::bind("0.0.0.0:0").ok()?;
    sock.set_read_timeout(Some(timeout)).ok()?;
    let addr = (server, 53u16).to_socket_addrs().ok()?.next()?;
    let packet = build_dns_query(domain);
    let start = std::time::Instant::now();
    sock.send_to(&packet, addr).ok()?;
    let mut buf = [0u8; 512];
    let (n, _) = sock.recv_from(&mut buf).ok()?;
    if n < 12 {
        return None;
    }
    Some(start.elapsed().as_millis() as u64)
}

/// Медиана из нескольких замеров (устойчивее среднего при выбросах).
fn median(mut v: Vec<u64>) -> Option<u64> {
    if v.is_empty() {
        return None;
    }
    v.sort_unstable();
    Some(v[v.len() / 2])
}

fn probe_addr(server: &str) -> Option<u64> {
    let mut samples = Vec::new();
    for _ in 0..3 {
        if let Some(ms) = query_once(server, "example.com", std::time::Duration::from_millis(1500)) {
            samples.push(ms);
        }
    }
    median(samples)
}

/// Замеряет время отклика DNS-серверов по всем провайдерам.
pub fn benchmark(ids: Option<Vec<String>>) -> Vec<DnsPing> {
    let selected: Vec<&DnsProvider> = match &ids {
        Some(list) if !list.is_empty() => PROVIDERS.iter().filter(|p| list.contains(&p.id.to_string())).collect(),
        _ => PROVIDERS.iter().collect(),
    };
    selected
        .into_iter()
        .map(|p| {
            let primary_ms = probe_addr(p.primary);
            let secondary_ms = probe_addr(p.secondary);
            let avg_ms = match (primary_ms, secondary_ms) {
                (Some(a), Some(b)) => Some((a + b) / 2),
                (Some(a), None) => Some(a),
                (None, Some(b)) => Some(b),
                _ => None,
            };
            DnsPing {
                id: p.id.to_string(),
                primary_ms,
                secondary_ms,
                avg_ms,
                error: if primary_ms.is_none() && secondary_ms.is_none() {
                    Some("нет ответа (UDP 53 закрыт/фильтруется)".into())
                } else {
                    None
                },
            }
        })
        .collect()
}

pub fn apply(data: &Path, id: &str, adapter: Option<&str>) -> Result<String, String> {
    let p = provider(id).ok_or("неизвестный DNS-провайдер")?;
    let script = data.join("logs").join(format!("dns_apply_{}.ps1", std::process::id()));
    let adapter_expr = match adapter.filter(|s| !s.trim().is_empty()) {
        Some(name) => crate::runner::ps_quote(name),
        None => "(Get-NetAdapter | Where-Object { $_.Status -eq 'Up' -and $_.HardwareInterface } | Select-Object -First 1 -ExpandProperty InterfaceAlias)".into(),
    };
    let body = format!(
        r#"{header}
$alias = {adapter}
if (-not $alias) {{ throw 'не найдено активное сетевое подключение' }}
$primary = '{primary}'
$secondary = '{secondary}'
$doh = '{doh}'
& netsh.exe dns add encryption server=$primary dohtemplate=$doh autoupgrade=yes udpfallback=no 2>$null | Out-Null
if ($LASTEXITCODE -ne 0) {{
  & netsh.exe dns set encryption server=$primary dohtemplate=$doh autoupgrade=yes udpfallback=no | Out-Null
  if ($LASTEXITCODE -ne 0) {{ throw "не удалось задать профиль DoH для $primary" }}
}}
& netsh.exe dns add encryption server=$secondary dohtemplate=$doh autoupgrade=yes udpfallback=no 2>$null | Out-Null
if ($LASTEXITCODE -ne 0) {{
  & netsh.exe dns set encryption server=$secondary dohtemplate=$doh autoupgrade=yes udpfallback=no | Out-Null
  if ($LASTEXITCODE -ne 0) {{ throw "не удалось задать профиль DoH для $secondary" }}
}}
Set-DnsClientServerAddress -InterfaceAlias $alias -ServerAddresses @($primary, $secondary)
Clear-DnsClientCache
Write-Output ("DNS: {name}; adapter: " + $alias)
"#,
        header = crate::runner::PS_HEADER,
        adapter = adapter_expr,
        primary = p.primary,
        secondary = p.secondary,
        doh = p.doh_template,
        name = p.name,
    );
    write_ps1(&script, &body)?;
    let code = run_elevated_script(&script)?;
    let _ = std::fs::remove_file(&script);
    if code != 0 {
        return Err(format!("применение DNS не удалось, код {}", code));
    }
    Ok(format!("{}: {} / {} + DoH (без UDP fallback)", p.name, p.primary, p.secondary))
}

pub fn reset(data: &Path, adapter: Option<&str>) -> Result<String, String> {
    let script = data.join("logs").join(format!("dns_reset_{}.ps1", std::process::id()));
    let adapter_expr = match adapter.filter(|s| !s.trim().is_empty()) {
        Some(name) => crate::runner::ps_quote(name),
        None => "(Get-NetAdapter | Where-Object { $_.Status -eq 'Up' -and $_.HardwareInterface } | Select-Object -First 1 -ExpandProperty InterfaceAlias)".into(),
    };
    let body = format!(
        r#"{header}
$alias = {adapter}
if (-not $alias) {{ throw 'не найдено активное сетевое подключение' }}
Set-DnsClientServerAddress -InterfaceAlias $alias -ResetServerAddresses
Clear-DnsClientCache
Write-Output ("DNS reset: " + $alias)
"#,
        header = crate::runner::PS_HEADER,
        adapter = adapter_expr,
    );
    write_ps1(&script, &body)?;
    let code = run_elevated_script(&script)?;
    let _ = std::fs::remove_file(&script);
    if code != 0 {
        return Err(format!("сброс DNS не удался, код {}", code));
    }
    Ok("DNS возвращён к автоматическим настройкам".into())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn providers_have_ipv4_and_doh() {
        assert!(providers().len() >= 8);
        for p in providers() {
            assert!(p.primary.parse::<std::net::Ipv4Addr>().is_ok());
            assert!(p.secondary.parse::<std::net::Ipv4Addr>().is_ok());
            assert!(p.doh_template.starts_with("https://"));
            assert!(!p.description.is_empty());
        }
    }

    #[test]
    fn finds_cloudflare_and_xbox() {
        assert_eq!(provider("cloudflare").unwrap().primary, "1.1.1.1");
        assert!(provider("comss").is_none());
        let x = provider("xbox-dns").unwrap();
        assert_eq!(x.primary, "111.88.96.50");
        assert_eq!(x.secondary, "111.88.96.51");
        assert_eq!(x.doh_template, "https://xbox-dns.ru/dns-query");
        assert!(provider("missing").is_none());
    }

    #[test]
    fn builds_valid_dns_query() {
        let q = build_dns_query("example.com");
        // 12 байт заголовка + 1+7 + 1+3 + 1(TLD-конец) + 4 (QTYPE/QCLASS)
        assert_eq!(q.len(), 12 + 8 + 4 + 1 + 4);
        assert_eq!(&q[12..13], &[7u8]); // длина "example"
        assert_eq!(&q[13..20], b"example");
    }
}

File: winforms/src/ZapretGui.Core/Net/Dns.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using ZapretGui.Core;
using ZapretGui.Core.Runtime;
using ZapretGui.Core.Util;

namespace ZapretGui.Core.Net
{
    public class DnsProvider
    {
        [JsonField("id")] public string Id;
        [JsonField("name")] public string Name;
        [JsonField("primary")] public string Primary;
        [JsonField("secondary")] public string Secondary;
        [JsonField("dohTemplate")] public string DohTemplate;
        /// <summary>Короткая подпись рядом с названием.</summary>
        [JsonField("note")] public string Note;
        /// <summary>Подробное описание (показывается в раскрывающемся блоке).</summary>
        [JsonField("description")] public string Description;

        public DnsProvider(string id, string name, string primary, string secondary,
            string dohTemplate, string note, string description)
        {
            Id = id;
            Name = name;
            Primary = primary;
            Secondary = secondary;
            DohTemplate = dohTemplate;
            Note = note;
            Description = description;
        }
    }

    public class DnsPing
    {
        [JsonField("id")] public string Id;
        [JsonField("primaryMs")] public long? PrimaryMs;
        [JsonField("secondaryMs")] public long? SecondaryMs;
        /// <summary>Среднее по обоим адресам (для сортировки/сравнения).</summary>
        [JsonField("avgMs")] public long? AvgMs;
        [JsonField("error")] public string Error;
    }

    /// <summary>
    /// DNS-провайдеры, применение/сброс DoH и бенчмарк. Порт dns.rs.
    /// </summary>
    public static class Dns
    {
        public static readonly DnsProvider[] Providers =
        {
            new DnsProvider("google", "Google Public DNS", "8.8.8.8", "8.8.4.4",
                "https://dns.google/dns-query", "крупный Anycast",
                "Публичный DNS Google. Очень стабильный и быстрый, без фильтрации. " +
                "Минус — принадлежит Google (логи). Хороший выбор по умолчанию."),
            new DnsProvider("cloudflare", "Cloudflare", "1.1.1.1", "1.0.0.1",
                "https://cloudflare-dns.com/dns-query", "быстрый, минимум логов",
                "Заявлен как самый быстрый публичный DNS, логов почти не ведёт. " +
                "Без фильтрации. Хорош для скорости и приватности."),
            new DnsProvider("adguard", "AdGuard DNS", "94.140.14.14", "94.140.15.15",
                "https://dns.adguard-dns.com/dns-query", "блокирует рекламу и трекеры",
                "Блокирует рекламу, трекеры и часть вредоносных доменов на уровне DNS. " +
                "Может резать и нужные сайты — если что-то не открывается, проверьте здесь."),
            new DnsProvider("xbox-dns", "XBox DNS", "111.88.96.50", "111.88.96.51",
                "https://xbox-dns.ru/dns-query", "Smart DNS: Xbox Live, ИИ, игры",
                "Smart DNS для доступа к сервисам с региональными ограничениями: " +
                "Xbox Live (ошибка 0x80a40401), игры Supercell (Brawl Stars, Clash of Clans), " +
                "ChatGPT/Gemini/Claude/Copilot, JetBrains/GitHub/Notion, Twitch/Spotify/xCloud. " +
                "Обычный трафик идёт напрямую, только гео-ограниченные — через шлюз. " +
                "Без регистрации, заявлены DoH/DoT и Zero-Log."),
            new DnsProvider("quad9", "Quad9", "9.9.9.9", "149.112.112.112",
                "https://dns.quad9.net/dns-query", "защита от вредоносных доменов",
                "Блокирует вредоносные и фишинговые домены (по базе угроз). " +
                "Без рекламных блокировок. Некоммерческий, ориентирован на безопасность."),
            new DnsProvider("opendns", "OpenDNS (Cisco)", "208.67.222.222", "208.67.220.220",
                "https://doh.opendns.com/dns-query", "стабильный, есть фильтры Cisco",
                "DNS от Cisco. Стабильный, с опциональной фильтрацией (по умолчанию — " +
                "базовая защита от фишинга). Хорош как надёжный резерв."),
            new DnsProvider("dns-sb", "DNS.SB", "185.222.222.222", "45.11.45.11",
                "https://doh.dns.sb/dns-query", "независимый, без логов",
                "Независимый публичный DNS, без логов и фильтрации. Хорошая альтернатива " +
                "большим корпоративным резолверам."),
            new DnsProvider("mullvad", "Mullvad DNS", "194.242.2.2", "194.242.2.3",
                "https://doh.mullvad.net/dns-query", "приватный, без логов",
                "DNS от Mullvad, известных своим отношением к приватности: " +
                "без логов, без фильтрации. Доступен без VPN-подписки."),
        };

        private const string DefaultAdapterExpr =
            "(Get-NetAdapter | Where-Object { $_.Status -eq 'Up' -and $_.HardwareInterface } | " +
            "Select-Object -First 1 -ExpandProperty InterfaceAlias)";

        public static DnsProvider Provider(string id)
        {
            foreach (var p in Providers) if (p.Id == id) return p;
            return null;
        }

        /// <summary>Строит минимальный DNS-запрос (A-запись) для домена. dns.rs:128</summary>
        public static byte[] BuildDnsQuery(string domain)
        {
            var q = new List<byte>(32);
            q.Add(0x12); q.Add(0x34); // ID
            q.Add(0x01); q.Add(0x00); // flags: standard query, RD
            q.Add(0x00); q.Add(0x01); // QDCOUNT
            q.Add(0x00); q.Add(0x00); // ANCOUNT
            q.Add(0x00); q.Add(0x00); // NSCOUNT
            q.Add(0x00); q.Add(0x00); // ARCOUNT
            foreach (var label in domain.Split('.'))
            {
                var bytes = Encoding.ASCII.GetBytes(label);
                q.Add((byte)bytes.Length);
                q.AddRange(bytes);
            }
            q.Add(0);
            q.Add(0x00); q.Add(0x01); // QTYPE A
            q.Add(0x00); q.Add(0x01); // QCLASS IN
            return q.ToArray();
        }

        /// <summary>Один UDP DNS-запрос: возвращает время отклика в мс (или null). dns.rs:147</summary>
        public static long? QueryOnce(string server, string domain, int timeoutMs)
        {
            try
            {
                var addresses = System.Net.Dns.GetHostAddresses(server);
                if (addresses.Length == 0) return null;
                var endPoint = new IPEndPoint(addresses[0], 53);
                var packet = BuildDnsQuery(domain);
                using (var sock = new UdpClient(new IPEndPoint(IPAddress.Any, 0)))
                {
                    sock.Client.ReceiveTimeout = timeoutMs;
                    var sw = Stopwatch.StartNew();
                    sock.Send(packet, packet.Length, endPoint);
                    var remoteEP = endPoint;
                    var buf = sock.Receive(ref remoteEP);
                    sw.Stop();
                    if (buf.Length < 12) return null;
                    return sw.ElapsedMilliseconds;
                }
            }
            catch { return null; }
        }

        /// <summary>Медиана из нескольких замеров (устойчивее среднего при выбросах). dns.rs:164</summary>
        public static long? Median(List<long> values)
        {
            if (values == null || values.Count == 0) return null;
            values.Sort();
            return values[values.Count / 2];
        }

        private static long? ProbeAddr(string server)
        {
            var samples = new List<long>();
            for (var i = 0; i < 3; i++)
            {
                var ms = QueryOnce(server, "example.com", 1500);
                if (ms.HasValue) samples.Add(ms.Value);
            }
            return Median(samples);
        }

        /// <summary>Замеряет время отклика DNS-серверов по выбранным провайдерам. dns.rs:183</summary>
        public static List<DnsPing> Benchmark(IList<string> ids)
        {
            var selected = new List<DnsProvider>();
            if (ids != null && ids.Count > 0)
            {
                foreach (var p in Providers) if (ids.Contains(p.Id)) selected.Add(p);
            }
            else
            {
                selected.AddRange(Providers);
            }

            var outList = new List<DnsPing>(selected.Count);
            foreach (var p in selected)
            {
                var primaryMs = ProbeAddr(p.Primary);
                var secondaryMs = ProbeAddr(p.Secondary);
                long? avgMs;
                if (primaryMs.HasValue && secondaryMs.HasValue) avgMs = (primaryMs.Value + secondaryMs.Value) / 2;
                else if (primaryMs.HasValue) avgMs = primaryMs.Value;
                else if (secondaryMs.HasValue) avgMs = secondaryMs.Value;
                else avgMs = null;

                outList.Add(new DnsPing
                {
                    Id = p.Id,
                    PrimaryMs = primaryMs,
                    SecondaryMs = secondaryMs,
                    AvgMs = avgMs,
                    Error = (!primaryMs.HasValue && !secondaryMs.HasValue)
                        ? "нет ответа (UDP 53 закрыт/фильтруется)"
                        : null,
                });
            }
            return outList;
        }

        /// <summary>Применяет DNS-провайдера (DoH через netsh). dns.rs:214</summary>
        public static Result<string> Apply(string dataDir, string id, string adapter)
        {
            var p = Provider(id);
            if (p == null) return Result<string>.Err("неизвестный DNS-провайдер");

            var script = Path.Combine(Path.Combine(dataDir, "logs"),
                "dns_apply_" + Process.GetCurrentProcess().Id + ".ps1");
            var adapterExpr = !string.IsNullOrWhiteSpace(adapter)
                ? Uac.PsQuote(adapter)
                : DefaultAdapterExpr;
            var body = Uac.PsHeader + "\n" +
                "$alias = " + adapterExpr + "\n" +
                "if (-not $alias) { throw 'не найдено активное сетевое подключение' }\n" +
                "$primary = '" + p.Primary + "'\n" +
                "$secondary = '" + p.Secondary + "'\n" +
                "$doh = '" + p.DohTemplate + "'\n" +
                "& netsh.exe dns add encryption server=$primary dohtemplate=$doh autoupgrade=yes udpfallback=no 2>$null | Out-Null\n" +
                "if ($LASTEXITCODE -ne 0) {\n" +
                "  & netsh.exe dns set encryption server=$primary dohtemplate=$doh autoupgrade=yes udpfallback=no | Out-Null\n" +
                "  if ($LASTEXITCODE -ne 0) { throw \"не удалось задать профиль DoH для $primary\" }\n" +
                "}\n" +
                "& netsh.exe dns add encryption server=$secondary dohtemplate=$doh autoupgrade=yes udpfallback=no 2>$null | Out-Null\n" +
                "if ($LASTEXITCODE -ne 0) {\n" +
                "  & netsh.exe dns set encryption server=$secondary dohtemplate=$doh autoupgrade=yes udpfallback=no | Out-Null\n" +
                "  if ($LASTEXITCODE -ne 0) { throw \"не удалось задать профиль DoH для $secondary\" }\n" +
                "}\n" +
                "Set-DnsClientServerAddress -InterfaceAlias $alias -ServerAddresses @($primary, $secondary)\n" +
                "Clear-DnsClientCache\n" +
                "Write-Output (\"DNS: " + p.Name + "; adapter: \" + $alias)\n";

            try { Uac.WritePs1(script, body); }
            catch (Exception e) { return Result<string>.Err(e.Message); }
            var code = Uac.RunElevatedScript(script);
            try { File.Delete(script); } catch { }
            if (code != 0) return Result<string>.Err("применение DNS не удалось, код " + code);
            return Result<string>.Ok(p.Name + ": " + p.Primary + " / " + p.Secondary + " + DoH (без UDP fallback)");
        }

        /// <summary>Сброс DNS к автоматическим настройкам. dns.rs:258</summary>
        public static Result<string> Reset(string dataDir, string adapter)
        {
            var script = Path.Combine(Path.Combine(dataDir, "logs"),
                "dns_reset_" + Process.GetCurrentProcess().Id + ".ps1");
            var adapterExpr = !string.IsNullOrWhiteSpace(adapter)
                ? Uac.PsQuote(adapter)
                : DefaultAdapterExpr;
            var body = Uac.PsHeader + "\n" +
                "$alias = " + adapterExpr + "\n" +
                "if (-not $alias) { throw 'не найдено активное сетевое подключение' }\n" +
                "Set-DnsClientServerAddress -InterfaceAlias $alias -ResetServerAddresses\n" +
                "Clear-DnsClientCache\n" +
                "Write-Output (\"DNS reset: \" + $alias)\n";

            try { Uac.WritePs1(script, body); }
            catch (Exception e) { return Result<string>.Err(e.Message); }
            var code = Uac.RunElevatedScript(script);
            try { File.Delete(script); } catch { }
            if (code != 0) return Result<string>.Err("сброс DNS не удался, код " + code);
            return Result<string>.Ok("DNS возвращён к автоматическим настройкам");
        }
    }
}
